src/mango/src/mango_cursor_view.erl

```diff
@@ -230,7 +231,13 @@ composite_prefix([Col | Rest], Ranges) ->
 % In the future we can look into doing a cached parallel
 % reduce view read on each index with the ranges to find
 % the one that has the fewest number of rows or something.
-choose_best_index(_DbName, IndexRanges) ->
+-type comparator() :: '$lt' | '$lte' | '$eq' | '$gte' | '$gt'.
+-type range() :: {comparator(), A, comparator(), A} | empty.
```

Comment:
What does `A` mean? If both types are `A`s do we imply they have to match? (I am not very familiar with the type syntax, so may be asking a silly question)

Reply:
Ah, ok. Thanks for explaining. There are parameterized types in Erlang https://www.erlang.org/doc/reference_manual/typespec.html#type-declarations-of-user-defined-types

It would be fantastic we if we could run a fast dialyzer check. Currently it takes a terribly long time to run it. Which is probably why it doesn't run in CI. There is also https://github.com/josefs/Gradualizer which might be fun to run as well.

Reply:
There was a plan to implement an incremental mode for dialyzer, but I don't know if it's released...

Reply:
That would be great to have @big-r81

Reply:
it seems that it's already [merged](https://github.com/erlang/otp/pull/5997) but not released yet... my `dialyzer` has no incremental option...

Reply:
@nickva I am not sure if introducing a parameter for the `range` type could help.  In the type of `mango_cursor_view:choose_best_index/1` there are lists of `range`s which should be homogeneous, i.e. applied to the same type (`[range(T)]` = `[range(T), range(T), ..]`).  But my understanding was that the ranges computed from the selector might potentially refer to different fields, therefore different types (`[range(T₁), range(T₂), .., range(Tₙ)]`).  Existential types can provide an answer for such situations: the type parameter gets enclosed within the definition of the type (`range`) and resolved locally on the spot.

Another solution could be an universal `range` type that contains all the possibly permitted variations, basically with the type parameter unrolled.  Something along these lines:

```erlang
-type r(T) :: {binary(), T, binary(), T}.
-type range() :: r(boolean()) | r(char()) | r(integer()) | ... | empty.
```

By the way, I think it might be beneficial to convert the binary strings for comparators to atoms, and then they could also be covered properly by types (and handled more efficiently, perhaps):

```erlang
-type comparator() :: lt | lte | eq | gte | gt.
-type r(T) :: {comparator(), T, comparator(), T}.
```

Although [I was not encouraged so much](https://learnyousomeerlang.com/types-or-lack-thereof#for-type-junkies) about getting too serious about types -- maybe I should just earn my own experiences on the topic and see how far it can be pushed in practice.

Reply:
In this case it might not help much, I just meant that in general we can have type parameters in type specs.

On a more practical level until it's quick and simple to run dialyzer locally and in a CI adding a lot of types is only helpful for documentation, but even then there is a danger they get out of sync with the actual types.

> By the way, I think it might be beneficial to convert the binary strings for comparators to atoms, and then they could also be covered properly by types (and handled more efficiently, perhaps):

There it looks like it's a mistake (sorry I should have caught it when reviewing the second time). Comparators are not binary, they are atoms. So even in this case we'd have an instance of a type spec being wrong and us not being able to validate or check it.